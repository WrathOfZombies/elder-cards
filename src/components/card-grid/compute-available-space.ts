import { Size } from "react-virtualized-auto-sizer";
import { FixedSizeGridProps } from "react-window";

import { MAX_CARD_HEIGHT, MAX_CARD_WIDTH } from "components/cards/trading-card";

const MIN_ROW_OR_COLUMN = 1;

export interface AvailableSpace extends Omit<FixedSizeGridProps, "children"> {
  columnCount: number;
  gutterSize: number;
}

/**
 * Compute the rows, columns and other properties based off of {@see Size} from the Autosize component.
 * This way we can limit the number of times the child re-renders as we only pass in, infrequently changing
 * primitives that can be memoized.
 * @returns An object of type {@see AvailableSpace}
 */
export const computeAvailableSpace = ({
  itemCount,
  width,
  height,
}: Size & { itemCount: number }): AvailableSpace => {
  // Divide the available width into columns
  // NO_OF_COLUMNS = AVAILABLE_WIDTH / ITEM_WIDTH
  // NO_OF_ROWS = ITEMS_COUNT / COLUMNS;
  const columnCount =
    Math.floor(width / MAX_CARD_WIDTH) ||
    MIN_ROW_OR_COLUMN; /* Min number of columns */

  const rowCount =
    Math.ceil(itemCount / columnCount) ||
    MIN_ROW_OR_COLUMN; /* Min number of rows */

  const renderedWidth = columnCount * MAX_CARD_WIDTH;

  // The gutter spacing can be calculated as:
  // GUTTER_SIZE = ((AVAILABLE_SPACE - MAX_SPACE_TAKEN_BY_COLMNS) / (TOTAL_COLUMS - 1)) / 2
  //
  // We add half the gutter size to create a "space around" effect
  // where the remaining available width in the UI is distributed among the items
  // horizontally
  // if availableSpace < renderedSpace (the case where there's no space for even 1 column) then gutters are 0
  const gutterSize =
    width > renderedWidth
      ? Math.abs((width - renderedWidth - 20) / (columnCount + 1))
      : 0;

  const columnWidth = MAX_CARD_WIDTH + gutterSize;
  const rowHeight = MAX_CARD_HEIGHT;

  return {
    width,
    height,
    rowCount,
    columnCount,
    columnWidth,
    rowHeight,
    gutterSize,
  };
};
