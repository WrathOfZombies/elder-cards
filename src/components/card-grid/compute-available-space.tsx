import { Size } from "react-virtualized-auto-sizer";
import { FixedSizeGridProps } from "react-window";

const MIN_ROW_OR_COLUMN = 1;
const ITEM_WIDTH = 300;
const ITEM_HEIGHT = 650;
const WINDOWS_SCROLLBAR = 20;

export interface IAvailableSpace extends Omit<FixedSizeGridProps, "children"> {
  columnCount: number;
  gutterSize: number;
}

/**
 * Compute the rows, columns and other properties based off of {@see Size} from the Autosize component.
 * This way we can limit the number of times the child re-renders as we only pass in, infrequently changing
 * primitives that can be memoized.
 * @returns An object of type {@see ISpaceDetectorProps}
 */
export const computeAvailableSpace = ({
  itemCount,
  width,
  height,
}: Size & { itemCount: number }): IAvailableSpace => {
  // Divide the available width into columns
  // NO_OF_COLUMNS = AVAILABLE_WIDTH / ITEM_WIDTH
  // NO_OF_ROWS = ITEMS_COUNT / COLUMNS;
  const columnCount =
    Math.floor(width / ITEM_WIDTH) ||
    MIN_ROW_OR_COLUMN; /* Min number of columns */

  const rowCount =
    Math.ceil(itemCount / columnCount) ||
    MIN_ROW_OR_COLUMN; /* Min number of rows */

  const renderedWidth = columnCount * ITEM_WIDTH;
  const renderedHeight = rowCount * ITEM_HEIGHT;

  // The gutter spacing can be calculated as:
  // GUTTER_SIZE = ((AVAILABLE_SPACE - MAX_SPACE_TAKEN_BY_COLMNS) / (TOTAL_COLUMS - 1)) / 2
  //
  // We add half the gutter size to create a "space around" effect
  // where the remaining available width in the UI is distributed among the items
  // horizontally
  // if availableSpace < renderedSpace (the case where there's no space for even 1 column) then gutters are 0
  const gutterSize =
    width > renderedWidth
      ? Math.abs(
          (width - renderedWidth - WINDOWS_SCROLLBAR) / (columnCount + 1)
        )
      : 0;
  // const gutterSize = 0;

  const columnWidth = ITEM_WIDTH + gutterSize;
  const rowHeight = ITEM_HEIGHT;

  return {
    width,
    height,
    rowCount,
    columnCount,
    columnWidth,
    rowHeight,
    gutterSize,
  };
};
