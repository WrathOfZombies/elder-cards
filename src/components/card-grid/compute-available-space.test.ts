import { computeAvailableSpace } from "./compute-available-space";

describe("Testing computeAvailableSpace", () => {
  const resolutions = [
    [240, 120],
    [360, 240],
    [640, 480],
    [1024, 720],
    [1280, 720],
    [1366, 768],
    [1920, 1080],
    [2560, 1440],
    [3820, 2160],
    [3440, 1440],
  ];

  const itemCounts = [0, 5, 43, 100];

  const cases = resolutions.flatMap(res =>
    itemCounts.map(items => [items, ...res])
  );

  test.each(cases)(
    "should compute the available space when itemCount is %p, width is %p and height is %p",
    (itemCount, width, height) => {
      const result = computeAvailableSpace({ itemCount, width, height });
      expect(result).toMatchSnapshot();
    }
  );
});
