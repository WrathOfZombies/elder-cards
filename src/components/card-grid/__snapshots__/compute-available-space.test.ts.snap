// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Testing computeAvailableSpace should compute the available space when itemCount is 0, width is 240 and height is 120 1`] = `
Object {
  "columnCount": 1,
  "columnWidth": 300,
  "gutterSize": 0,
  "height": 120,
  "rowCount": 1,
  "rowHeight": 670,
  "width": 240,
}
`;

exports[`Testing computeAvailableSpace should compute the available space when itemCount is 0, width is 360 and height is 240 1`] = `
Object {
  "columnCount": 1,
  "columnWidth": 320,
  "gutterSize": 20,
  "height": 240,
  "rowCount": 1,
  "rowHeight": 670,
  "width": 360,
}
`;

exports[`Testing computeAvailableSpace should compute the available space when itemCount is 0, width is 640 and height is 480 1`] = `
Object {
  "columnCount": 2,
  "columnWidth": 306.6666666666667,
  "gutterSize": 6.666666666666667,
  "height": 480,
  "rowCount": 1,
  "rowHeight": 670,
  "width": 640,
}
`;

exports[`Testing computeAvailableSpace should compute the available space when itemCount is 0, width is 1024 and height is 720 1`] = `
Object {
  "columnCount": 3,
  "columnWidth": 326,
  "gutterSize": 26,
  "height": 720,
  "rowCount": 1,
  "rowHeight": 670,
  "width": 1024,
}
`;

exports[`Testing computeAvailableSpace should compute the available space when itemCount is 0, width is 1280 and height is 720 1`] = `
Object {
  "columnCount": 4,
  "columnWidth": 312,
  "gutterSize": 12,
  "height": 720,
  "rowCount": 1,
  "rowHeight": 670,
  "width": 1280,
}
`;

exports[`Testing computeAvailableSpace should compute the available space when itemCount is 0, width is 1366 and height is 768 1`] = `
Object {
  "columnCount": 4,
  "columnWidth": 329.2,
  "gutterSize": 29.2,
  "height": 768,
  "rowCount": 1,
  "rowHeight": 670,
  "width": 1366,
}
`;

exports[`Testing computeAvailableSpace should compute the available space when itemCount is 0, width is 1920 and height is 1080 1`] = `
Object {
  "columnCount": 6,
  "columnWidth": 314.2857142857143,
  "gutterSize": 14.285714285714286,
  "height": 1080,
  "rowCount": 1,
  "rowHeight": 670,
  "width": 1920,
}
`;

exports[`Testing computeAvailableSpace should compute the available space when itemCount is 0, width is 2560 and height is 1440 1`] = `
Object {
  "columnCount": 8,
  "columnWidth": 315.55555555555554,
  "gutterSize": 15.555555555555555,
  "height": 1440,
  "rowCount": 1,
  "rowHeight": 670,
  "width": 2560,
}
`;

exports[`Testing computeAvailableSpace should compute the available space when itemCount is 0, width is 3440 and height is 1440 1`] = `
Object {
  "columnCount": 11,
  "columnWidth": 310,
  "gutterSize": 10,
  "height": 1440,
  "rowCount": 1,
  "rowHeight": 670,
  "width": 3440,
}
`;

exports[`Testing computeAvailableSpace should compute the available space when itemCount is 0, width is 3820 and height is 2160 1`] = `
Object {
  "columnCount": 12,
  "columnWidth": 315.38461538461536,
  "gutterSize": 15.384615384615385,
  "height": 2160,
  "rowCount": 1,
  "rowHeight": 670,
  "width": 3820,
}
`;

exports[`Testing computeAvailableSpace should compute the available space when itemCount is 5, width is 240 and height is 120 1`] = `
Object {
  "columnCount": 1,
  "columnWidth": 300,
  "gutterSize": 0,
  "height": 120,
  "rowCount": 5,
  "rowHeight": 670,
  "width": 240,
}
`;

exports[`Testing computeAvailableSpace should compute the available space when itemCount is 5, width is 360 and height is 240 1`] = `
Object {
  "columnCount": 1,
  "columnWidth": 320,
  "gutterSize": 20,
  "height": 240,
  "rowCount": 5,
  "rowHeight": 670,
  "width": 360,
}
`;

exports[`Testing computeAvailableSpace should compute the available space when itemCount is 5, width is 640 and height is 480 1`] = `
Object {
  "columnCount": 2,
  "columnWidth": 306.6666666666667,
  "gutterSize": 6.666666666666667,
  "height": 480,
  "rowCount": 3,
  "rowHeight": 670,
  "width": 640,
}
`;

exports[`Testing computeAvailableSpace should compute the available space when itemCount is 5, width is 1024 and height is 720 1`] = `
Object {
  "columnCount": 3,
  "columnWidth": 326,
  "gutterSize": 26,
  "height": 720,
  "rowCount": 2,
  "rowHeight": 670,
  "width": 1024,
}
`;

exports[`Testing computeAvailableSpace should compute the available space when itemCount is 5, width is 1280 and height is 720 1`] = `
Object {
  "columnCount": 4,
  "columnWidth": 312,
  "gutterSize": 12,
  "height": 720,
  "rowCount": 2,
  "rowHeight": 670,
  "width": 1280,
}
`;

exports[`Testing computeAvailableSpace should compute the available space when itemCount is 5, width is 1366 and height is 768 1`] = `
Object {
  "columnCount": 4,
  "columnWidth": 329.2,
  "gutterSize": 29.2,
  "height": 768,
  "rowCount": 2,
  "rowHeight": 670,
  "width": 1366,
}
`;

exports[`Testing computeAvailableSpace should compute the available space when itemCount is 5, width is 1920 and height is 1080 1`] = `
Object {
  "columnCount": 6,
  "columnWidth": 314.2857142857143,
  "gutterSize": 14.285714285714286,
  "height": 1080,
  "rowCount": 1,
  "rowHeight": 670,
  "width": 1920,
}
`;

exports[`Testing computeAvailableSpace should compute the available space when itemCount is 5, width is 2560 and height is 1440 1`] = `
Object {
  "columnCount": 8,
  "columnWidth": 315.55555555555554,
  "gutterSize": 15.555555555555555,
  "height": 1440,
  "rowCount": 1,
  "rowHeight": 670,
  "width": 2560,
}
`;

exports[`Testing computeAvailableSpace should compute the available space when itemCount is 5, width is 3440 and height is 1440 1`] = `
Object {
  "columnCount": 11,
  "columnWidth": 310,
  "gutterSize": 10,
  "height": 1440,
  "rowCount": 1,
  "rowHeight": 670,
  "width": 3440,
}
`;

exports[`Testing computeAvailableSpace should compute the available space when itemCount is 5, width is 3820 and height is 2160 1`] = `
Object {
  "columnCount": 12,
  "columnWidth": 315.38461538461536,
  "gutterSize": 15.384615384615385,
  "height": 2160,
  "rowCount": 1,
  "rowHeight": 670,
  "width": 3820,
}
`;

exports[`Testing computeAvailableSpace should compute the available space when itemCount is 43, width is 240 and height is 120 1`] = `
Object {
  "columnCount": 1,
  "columnWidth": 300,
  "gutterSize": 0,
  "height": 120,
  "rowCount": 43,
  "rowHeight": 670,
  "width": 240,
}
`;

exports[`Testing computeAvailableSpace should compute the available space when itemCount is 43, width is 360 and height is 240 1`] = `
Object {
  "columnCount": 1,
  "columnWidth": 320,
  "gutterSize": 20,
  "height": 240,
  "rowCount": 43,
  "rowHeight": 670,
  "width": 360,
}
`;

exports[`Testing computeAvailableSpace should compute the available space when itemCount is 43, width is 640 and height is 480 1`] = `
Object {
  "columnCount": 2,
  "columnWidth": 306.6666666666667,
  "gutterSize": 6.666666666666667,
  "height": 480,
  "rowCount": 22,
  "rowHeight": 670,
  "width": 640,
}
`;

exports[`Testing computeAvailableSpace should compute the available space when itemCount is 43, width is 1024 and height is 720 1`] = `
Object {
  "columnCount": 3,
  "columnWidth": 326,
  "gutterSize": 26,
  "height": 720,
  "rowCount": 15,
  "rowHeight": 670,
  "width": 1024,
}
`;

exports[`Testing computeAvailableSpace should compute the available space when itemCount is 43, width is 1280 and height is 720 1`] = `
Object {
  "columnCount": 4,
  "columnWidth": 312,
  "gutterSize": 12,
  "height": 720,
  "rowCount": 11,
  "rowHeight": 670,
  "width": 1280,
}
`;

exports[`Testing computeAvailableSpace should compute the available space when itemCount is 43, width is 1366 and height is 768 1`] = `
Object {
  "columnCount": 4,
  "columnWidth": 329.2,
  "gutterSize": 29.2,
  "height": 768,
  "rowCount": 11,
  "rowHeight": 670,
  "width": 1366,
}
`;

exports[`Testing computeAvailableSpace should compute the available space when itemCount is 43, width is 1920 and height is 1080 1`] = `
Object {
  "columnCount": 6,
  "columnWidth": 314.2857142857143,
  "gutterSize": 14.285714285714286,
  "height": 1080,
  "rowCount": 8,
  "rowHeight": 670,
  "width": 1920,
}
`;

exports[`Testing computeAvailableSpace should compute the available space when itemCount is 43, width is 2560 and height is 1440 1`] = `
Object {
  "columnCount": 8,
  "columnWidth": 315.55555555555554,
  "gutterSize": 15.555555555555555,
  "height": 1440,
  "rowCount": 6,
  "rowHeight": 670,
  "width": 2560,
}
`;

exports[`Testing computeAvailableSpace should compute the available space when itemCount is 43, width is 3440 and height is 1440 1`] = `
Object {
  "columnCount": 11,
  "columnWidth": 310,
  "gutterSize": 10,
  "height": 1440,
  "rowCount": 4,
  "rowHeight": 670,
  "width": 3440,
}
`;

exports[`Testing computeAvailableSpace should compute the available space when itemCount is 43, width is 3820 and height is 2160 1`] = `
Object {
  "columnCount": 12,
  "columnWidth": 315.38461538461536,
  "gutterSize": 15.384615384615385,
  "height": 2160,
  "rowCount": 4,
  "rowHeight": 670,
  "width": 3820,
}
`;

exports[`Testing computeAvailableSpace should compute the available space when itemCount is 100, width is 240 and height is 120 1`] = `
Object {
  "columnCount": 1,
  "columnWidth": 300,
  "gutterSize": 0,
  "height": 120,
  "rowCount": 100,
  "rowHeight": 670,
  "width": 240,
}
`;

exports[`Testing computeAvailableSpace should compute the available space when itemCount is 100, width is 360 and height is 240 1`] = `
Object {
  "columnCount": 1,
  "columnWidth": 320,
  "gutterSize": 20,
  "height": 240,
  "rowCount": 100,
  "rowHeight": 670,
  "width": 360,
}
`;

exports[`Testing computeAvailableSpace should compute the available space when itemCount is 100, width is 640 and height is 480 1`] = `
Object {
  "columnCount": 2,
  "columnWidth": 306.6666666666667,
  "gutterSize": 6.666666666666667,
  "height": 480,
  "rowCount": 50,
  "rowHeight": 670,
  "width": 640,
}
`;

exports[`Testing computeAvailableSpace should compute the available space when itemCount is 100, width is 1024 and height is 720 1`] = `
Object {
  "columnCount": 3,
  "columnWidth": 326,
  "gutterSize": 26,
  "height": 720,
  "rowCount": 34,
  "rowHeight": 670,
  "width": 1024,
}
`;

exports[`Testing computeAvailableSpace should compute the available space when itemCount is 100, width is 1280 and height is 720 1`] = `
Object {
  "columnCount": 4,
  "columnWidth": 312,
  "gutterSize": 12,
  "height": 720,
  "rowCount": 25,
  "rowHeight": 670,
  "width": 1280,
}
`;

exports[`Testing computeAvailableSpace should compute the available space when itemCount is 100, width is 1366 and height is 768 1`] = `
Object {
  "columnCount": 4,
  "columnWidth": 329.2,
  "gutterSize": 29.2,
  "height": 768,
  "rowCount": 25,
  "rowHeight": 670,
  "width": 1366,
}
`;

exports[`Testing computeAvailableSpace should compute the available space when itemCount is 100, width is 1920 and height is 1080 1`] = `
Object {
  "columnCount": 6,
  "columnWidth": 314.2857142857143,
  "gutterSize": 14.285714285714286,
  "height": 1080,
  "rowCount": 17,
  "rowHeight": 670,
  "width": 1920,
}
`;

exports[`Testing computeAvailableSpace should compute the available space when itemCount is 100, width is 2560 and height is 1440 1`] = `
Object {
  "columnCount": 8,
  "columnWidth": 315.55555555555554,
  "gutterSize": 15.555555555555555,
  "height": 1440,
  "rowCount": 13,
  "rowHeight": 670,
  "width": 2560,
}
`;

exports[`Testing computeAvailableSpace should compute the available space when itemCount is 100, width is 3440 and height is 1440 1`] = `
Object {
  "columnCount": 11,
  "columnWidth": 310,
  "gutterSize": 10,
  "height": 1440,
  "rowCount": 10,
  "rowHeight": 670,
  "width": 3440,
}
`;

exports[`Testing computeAvailableSpace should compute the available space when itemCount is 100, width is 3820 and height is 2160 1`] = `
Object {
  "columnCount": 12,
  "columnWidth": 315.38461538461536,
  "gutterSize": 15.384615384615385,
  "height": 2160,
  "rowCount": 9,
  "rowHeight": 670,
  "width": 3820,
}
`;
